RFC: simplify sstable reader

wmitros opened the pull request:
This patch simplifies the sstable reader by getting rid of the state that we needed to remember in the sstable reader and running in a coroutine instead(as suggested in #7952).
I'm posting the RFC mainly to get an opinion on the synchronization approach that was used, I still need to fix the style of the patch, and add the same change for mx sstables reader.
We've decided with @haaawk that instead of filling a queue, and waiting for `fill_buffer()` to move `mutation_fragments` from the queue to the buffer, we can limit the `mutation_fragments` generation to the `fill_buffer()` duration, when we can skip the queue altogether and fill the buffer directly.
We're reading and consuming the sstable in a coroutine `producer_main()`, in which we wait on a semaphore for a `fill_buffer()` call. In the producer coroutine we wait for buffers from the sstable input_stream, and after consuming enough, we wait on the semaphore again, resuming caller of fill_buffer. The errors thrown in the coroutine are returned through the semaphore using broken().
This change does not improve the efficiency of sstable reading, but after some more style improvements, it should make following the production easier

haaawk commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,277 +144,248 @@ private:
-        }
+    future<> await_production() {
+        _production_sem.signal();
+        co_return co_await _production_sem.wait();
```
It might be safer to make two semaphores one for data_consume_rows_context -> producer_main communication and another for producer_main->data_consume_rows_context. Just to make sure there are no weird races between them.

avikivity replied:
We want to reduce the number of semaphores, not increase them.

avikivity replied:
(there should be no races in this code, since there is no real concurrency)

haaawk replied:
What about producer_main coroutine being started in the constructor of data_consume_rows_context and fill_buffer being called before that coroutine reaches _production_sem.wait()?

haaawk replied:
Maybe that's impossible - I really don't have any intuition about coroutines.

avikivity replied:
I'm not saying if it's possible or not, I'm saying that the logical path has no concurrency. We read data from disk, wait for it, parse in memory, wait if the buffer is full, resume, parse some more, run out of data, read from disk, rinse and repeat. Nothing happens in parallel. So if there are races, this code introduces them. Rather than fixing with an expensive semaphore, it should not introduce them in the first place.

haaawk replied:
Fair enough.

denesb replied:
Couldn't we convert all this to a pull model with `co_yield`? All the complication rises from the currently used push model, where the reader starts a fiber, which pushes fragments into the reader's buffer, creating a circle (R=reader, P=parser):

reader (R) -> read() (P) -> push fragment (P) -> is buffer full (R) -> yes ? stop/continue (P)

With a generator we could have a simple loop in the reader's `fill_buffer()` (fantasy pseudo code follows):

```C++
while (!_is_buffer_full) {
    push_mutation_fragment(co_await _parser.get_next_fragment());
}
```
Internally, `get_next_fragment()`, uses `co_yield` to return the last parsed fragment and to suspend its state until the next fragment is requested.

tgrabiec replied:
I think co_yield will be inefficient because it would context switch on each fragment

denesb replied:
Is it really a context switch though? You are basically just calling a function for each fragment.

avikivity replied:
I think the cost is minor. As long as we are not yielding to the consumer (and just buffering) there are few extra instructions added. Now that perf_fast_forward reports instructions per fragment this can be easily measured.

tgrabiec replied:
If co_yield buffers, then the context switch is amortized. 

It still seems problematic. If the parser buffers, it has to over-parse. The reader could stop earlier.

I don't think it simplifies things. I see how it makes things more complicated. The parser knows the type of the fragment. If we produce mutation_fragment, this information is erased and the reader has to figure it out again. The same logic which would live in the parser's handler will also have to live in the reader's loop.

avikivity replied:
> If co_yield buffers, then the context switch is amortized.
> 
> It still seems problematic. If the parser buffers, it has to over-parse. The reader could stop earlier.
> 

The idea is to have the same behavior as now in terms of buffering, just with a nicer interface.

`data_consume_rows_context_m::_state` is replaced with the coroutine remembering the instruction pointer across suspends.
switch/goto are replaced with normal nested loops
`co_yield ...` calls `operator co_await`, and on the return object of that it calls `await_ready()`. We can arrange for that to return false if no preemption needs to happen and if the buffer is not full; this should translate directly to the existing `if` checks.


> I don't think it simplifies things. I see how it makes things more complicated. The parser knows the type of the fragment. If we produce mutation_fragment, this information is erased and the reader has to figure it out again. The same logic which would live in the parser's handler will also have to live in the reader's loop.

The information is not erased, it remains in the instruction pointer. Any local state (which is now in member variables) is also retained.

avikivity commented:
Note k/l is old, there is no point in converting it.

avikivity commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,277 +144,248 @@ private:
+        while (rs != data_consumer::read_status::ready) {
+            assert(_data.size() == 0);
+            co_await read_nonempty();
+            rs = primitive_consumer::consume(_data);
```
I was afraid of this. A coroutine always allocates, so each produce_64() is an allocation.

In the existing code, reading some primitive has some overhead even in the common case where the data is ready in the buffer, but it's just some checks. Now we have to allocate coroutine state each time.

Did you measure performance? Not sure what would make a good test for this.

avikivity replied:
Perhaps the primitives can be converted to continuations. But we'll still see overhead from futures, which are too expensive for such simple calculations.

raphaelsc replied:
> 
> Did you measure performance? Not sure what would make a good test for this.

perhaps test/perf/perf_sstable after some tweaking

avikivity replied:
I think if coroutines can be used here, they would be custom coroutines, not seastar coroutines. We'd replace proceed::no with `co_yield proceed::no`. The coroutine return value would be a custom wrapper around the handle.

It's still problematic. The entire parser would have to be one giant function (so we have one allocation per parse).

avikivity replied:
(it would replace the big switch and gotos with standard control flow, but that's about it)

tgrabiec replied:
For testing you can use perf_fast_forward where it scans multi partition data set.

@avikivity Why do you want to use co_yield with proceed::no?

avikivity replied:
I want to replace the switch/goto based control flow with nested loops for clarity. But I'm not willing to give up performance for it, and the switch thing is pretty much optimal.

This is likely quite a hard task.

avikivity commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,277 +144,248 @@ private:
+    }
+    future<> wait() {
+        finish_reading();
+        _production_sem.signal();
```
This is penalizing the common case where everything that needs to be read can be consumed in one, without ever yielding. This function will always go through the scheduler. perf_fast_forward (in the simplest case where it never fast-forwards) will report a huge increase in tasks and allocations here.

(let's also make it report instructions)

tgrabiec commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,277 +144,248 @@ private:
+            rs = primitive_consumer::consume(_data);
         }
-#endif
-        sstlog.trace("data_consume_row_context {}: state={}, size={}", fmt::ptr(this), static_cast<int>(_state), data.size());
```
Please keep the trace logging in some way.

wmitros replied:
I think I deleted this by mistake when clearing up my temporary debug, will remember to keep it in the final version

tgrabiec commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,277 +144,248 @@ private:
-            if (ret == row_consumer::proceed::no) {
-                return row_consumer::proceed::no;
+            skip_row = false;
+            while (true) { // atom loop
```
Not if it's a coroutine though, due to allocation.

wmitros replied:
I suspect it would be a coroutine, but we're moving away from this solution anyway

denesb commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,277 +144,248 @@ private:
-                auto v = fragmented_temporary_buffer::view(_val_fragmented);
-                read_fragmented(v, 4, reinterpret_cast<bytes::value_type*>(_val.get_write()));
+                co_await produce_short_length_bytes(_key);
+                co_await produce_32();
```
I agree, this feels like reading assembly, where each command reads its parameters from and writes its result into some register, and you are supposed to know which where.

wmitros replied:
Right, I was planning to change it in the final version, but considering that this is how we're reading the primitives currently, I left it as it is in the rfc

denesb commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,277 +144,248 @@ private:
+                    _deleted = false;
+                    _counter = true;
+                    co_await produce_64();
+                    // _timestamp_of_last_deletion = _u64;
```
?

avikivity commented:
Note perf_fast_forward now reports instructions per fragment.

wmitros commented:
Sorry for the late reply, I'll share my new results in this single comment because there are some longer comment threads with overlaying ideas, so I'd be repeating myself answering each one.

I tested the new reader and some modifications of it with perf_sstable (unfortunately I changed the older reader, and perf_fast_forward didn't work with it). I tested sequential read speed and got:

- ~234k partitions/sec for the original reader
- ~114k partitions/sec for the submitted version
- ~198k partitions/sec for the submitted version with `produce_*` methods changed to continuations
- ~230k partitions/sec for @avikivity's suggestion with `co_yield proceed::no

If we want to keep up current speed, the only viable options seem to be slightly improving the solution with `co_yield proceed::no`, or possibly @denesb's idea with co_yielding mutation_fragments, which would require some more work.

The issue with these two options is that we would need to implement our own generator class, which would be about twice as big as coroutine_traits_base<> in seastar just for this use, and I'm not sure whether the sstable reader methods could be simplified enough to outweigh this additional logic.

Maybe a solution to that would be adding some general implementation for generators to seastar, but I'm not sure how would it look like considering how much coroutines with co_yield can offer.

avikivity commented:
> 
> The issue with these two options is that we would need to implement our own generator class, which would be about twice as big as coroutine_traits_base<> in seastar just for this use, and I'm not sure whether the sstable reader methods could be simplified enough to outweigh this additional logic.

If we can convert the switch/goto combination to structured control flow with similar performance it would be a win IMO, worth the coroutine_traits (but with local use only, it can't reach into seastar::task etc, so it can't be future/promise based).

> 
> Maybe a solution to that would be adding some general implementation for generators to seastar, but I'm not sure how would it look like considering how much coroutines with co_yield can offer.

I want to add a stream coroutine adapter to Seastar. The problem is that there is a lot of tradeoffs to be made (what container to use, how to decide when the container is full enough to yield). So it is hard to design in a way that is easy to use, yet leaves these decisions in the hand of users.

Strawman:

```c++
template <typename T, stream_options options = default_stream_options<T>>
class stream {
       // methods to push and pop, or perhaps these should be separated into two classes
};

template <typename T>
struct default_stream_options {
    using container = circular_buffer<T>;
    static bool buffer_full(const container& c) { return c.size() > 100; }
};

stream<mutation_fragment> read_sstable(...) {
    ...
    auto a_fragment = ...;
    co_yield std::move(a_fragment);
    ...
}
```

wmitros commented:
Before trying to merge, I've decided to ask for some more comments to the unfinished version, because there's already much to review, and I'm not sure of some changes.

The most recent version should be equivalent to the original version in the sense that both perform the same set of instructions in the same order (except the instructions considering state, returning vs co_yielding, etc.).
The main remaining change I'm planning to add is replacing private variables of the `data_consume_rows_*` classes with local variables in `do_process_state` so that they are destroyed when out of scope, and we don't have to clear them after each use.

One issue I have with the uploaded version is the `SEASTAR_INTERNAL_COROUTINE_NAMESPACE::coroutine_traits<proceed_generator,...>` class. We have to define a `promise_type` as its nested class, so that `proceed_generator` can be a return type of a coroutine, but `SEASTAR_INTERNAL_COROUTINE_NAMESPACE` is `std` or `std::experimental`. I've added the definition to `sstables/proceed_generator.hh`, but I'm not sure we can do that, and even if we do, should it be at that location?

Another issue is how much are we against the switch-case with states. In the mx parser I've removed all but 5 cases, and it doesn't seem like the control with loops and ifs can get simpler than this switch.

The last thing I'd like to ask is whether the `non_consuming()` method does mean anything specific? I was expecting that it would behave in one of 2 ways: either return true for states in that perform some actions and always return without reading any new data, or states that can return without reading at least some times, but neither description looks accurate.

wmitros commented:
Also, the commits are not cleaned up (at least 1 or 2 things are added and later removed, and the original code that's been commented out has been removed only in some of the last commits), so it may be better to look at the sum of changes

wmitros commented:
> The main remaining change I'm planning to add is replacing private variables of the `data_consume_rows_*` classes with local variables in `do_process_state` so that they are destroyed when out of scope, and we don't have to clear them after each use.

After a little trial it seems like moving variables from the class to the generator doesn't really simplify anything, so I think I'll skip it

> 
> One issue I have with the uploaded version is the `SEASTAR_INTERNAL_COROUTINE_NAMESPACE::coroutine_traits<proceed_generator,...>` class. We have to define a `promise_type` as its nested class, so that `proceed_generator` can be a return type of a coroutine, but `SEASTAR_INTERNAL_COROUTINE_NAMESPACE` is `std` or `std::experimental`. I've added the definition to `sstables/proceed_generator.hh`, but I'm not sure we can do that, and even if we do, should it be at that location?

I've seen that we are implementing some things in namespace std (mostly hash and operator<<), but I'd still like to get an confirmation on the approach. I've seen you've been working on coroutine namespaces, so @avikivity could you share an opinion about this?

> 
> Another issue is how much are we against the switch-case with states. In the mx parser I've removed all but 5 cases, and it doesn't seem like the control with loops and ifs can get simpler than this switch.
> 
> The last thing I'd like to ask is whether the `non_consuming()` method does mean anything specific? I was expecting that it would behave in one of 2 ways: either return true for states in that perform some actions and always return without reading any new data, or states that can return without reading at least some times, but neither description looks accurate.

Although there is some explanation about this in `consumer.hh`, I'm still not sure which are exactly the cases when this should be true, but seeing that it's used in only one place: `while (data || (!primitive_consumer::active() && non_consuming()))`, and most of the possible places we're returning (yielding) are when primitive_consumer::active(), I managed to reduce it to almost nothing, without changing its results that are being checked

avikivity commented:
> > The main remaining change I'm planning to add is replacing private variables of the `data_consume_rows_*` classes with local variables in `do_process_state` so that they are destroyed when out of scope, and we don't have to clear them after each use.
> 
> After a little trial it seems like moving variables from the class to the generator doesn't really simplify anything, so I think I'll skip it
> 
> > One issue I have with the uploaded version is the `SEASTAR_INTERNAL_COROUTINE_NAMESPACE::coroutine_traits<proceed_generator,...>` class. We have to define a 

Don't use SEASTAR_INTERNAL things. Seastar has to support multiple compilers/standards, we don't, so just pick the one that matches our compiler.

> `promise_type` as its nested class, so that `proceed_generator` can be a return type of a coroutine, but `SEASTAR_INTERNAL_COROUTINE_NAMESPACE` is `std` or `std::experimental`. I've added the definition to `sstables/proceed_generator.hh`, but I'm not sure we can do that, and even if we do, should it be at that location?

Just use std or std::experimental, whichever works.

> 
> I've seen that we are implementing some things in namespace std (mostly hash and operator<<), but I'd still like to get an confirmation on the approach. I've seen you've been working on coroutine namespaces, so @avikivity could you share an opinion about this?

std::hash and std::coroutine_traits are extension points, you are expected to specialize them. There is no other way to use coroutines.

> 
> > Another issue is how much are we against the switch-case with states. In the mx parser I've removed all but 5 cases, and it doesn't seem like the control with loops and ifs can get simpler than this switch.
> > The last thing I'd like to ask is whether the `non_consuming()` method does mean anything specific? I was expecting that it would behave in one of 2 ways: either return true for states in that perform some actions and always return without reading any new data, or states that can return without reading at least some times, but neither description looks accurate.
> 
> Although there is some explanation about this in `consumer.hh`, I'm still not sure which are exactly the cases when this should be true, but seeing that it's used in only one place: `while (data || (!primitive_consumer::active() && non_consuming()))`, and most of the possible places we're returning (yielding) are when primitive_consumer::active(), I managed to reduce it to almost nothing, without changing its results that are being checked

What does the performance look like?

I'm especially interested in instructions per row (you can make it more stable by running over tmpfs).

wmitros commented:
I've run perf_fast_forward on scylla with and without the change on /tmp, and from the results I've got, the instructions per fragment were higher by up to 4.8% in some cases. If we give each case the same weight, the average difference was 2.1%
[new_result.txt](https://github.com/scylladb/scylla/files/6631840/new_result.txt)
[old_result.txt](https://github.com/scylladb/scylla/files/6631841/old_result.txt)
I'm not sure if we have more tests that count instructions

avikivity commented:
The simplification is considerable but the performance loss is not trivial.

I see some states in mx reader remain. Why were they not converted? I'm wondering if converting them would recover performance or increase the loss, I don't have intuition on this.

The most important case is when skips=0 (with/without an offset). If we could have the same or similar performance there, it would be easier to decide.

wmitros commented:
I'm back from my exams, so I can finally get back to finishing this.


> The simplification is considerable but the performance loss is not trivial.
> 
> I see some states in mx reader remain. Why were they not converted? I'm wondering if converting them would recover performance or increase the loss, I don't have intuition on this.

We still need to have some states for ```verify_end_state()```, so I decided not to create new ones for this, but use the existing ones instead. Also, you can enter each of these states from multiple places, and to perform this without adding new coroutines (so new allocs) you'd have to use ifs/whiles spanning hundreds of lines, which imo made the code more complex instead of simplifying it.
Considering we're already using them, I replaced these states with ```goto```s and labels, and it helped in both performance and readability.
We may also need states for the ```reset()``` method, if we decide not to reconstruct the generator object after each ```reset()``` call (though in my testing it didn't bring any measurable performance boost)

> The most important case is when skips=0 (with/without an offset). If we could have the same or similar performance there, it would be easier to decide.

For some reason, the new code performs worst when we're starting with a big offset or when we're doing large skips. I couldn't find where in the changes the size of an offset could make a difference, but that's what the tests consistently showed.
The code also performed better when we were reading many rows (probably as expected, because we only need to construct the coroutine generator once to read many fragments).
When these two factors were decent, the change was less than 2% most of the times

avikivity commented:
> I'm back from my exams, so I can finally get back to finishing this.
> 
> > The simplification is considerable but the performance loss is not trivial.
> > I see some states in mx reader remain. Why were they not converted? I'm wondering if converting them would recover performance or increase the loss, I don't have intuition on this.
> 
> We still need to have some states for `verify_end_state()`, so I decided not to create new ones for this, but use the existing ones instead. Also, you can enter each of these states from multiple places, and to perform this without adding new coroutines (so new allocs) you'd have to use ifs/whiles spanning hundreds of lines, which imo made the code more complex instead of simplifying it.
> Considering we're already using them, I replaced these states with `goto`s and labels, and it helped in both performance and readability.
> We may also need states for the `reset()` method, if we decide not to reconstruct the generator object after each `reset()` call (though in my testing it didn't bring any measurable performance boost)
> 
> > The most important case is when skips=0 (with/without an offset). If we could have the same or similar performance there, it would be easier to decide.
> 
> For some reason, the new code performs worst when we're starting with a big offset or when we're doing large skips. I couldn't find where in the changes the size of an offset could make a difference, but that's what the tests consistently showed.
> The code also performed better when we were reading many rows (probably as expected, because we only need to construct the coroutine generator once to read many fragments).
> When these two factors were decent, the change was less than 2% most of the times

What's the worst case degradation? Both the case, and how much we degrade.

wmitros commented:
The biggest increase in instructions per fragment is in the case ```large-partition-select-few-rows``` (where we select between 1 and 16 rows, with jumps in-between), but after changing the switch statement to ```goto```s, and adding a state for resetting the generator without reconstructing, we still have up to 4% more.
There are 3 sets of ```large-partition-slicing``` tests where we read some rows after a big offset, the loss is similar among them, but the highest is in ```large-partition-slicing-clustering-keys```, when reading 1 row we also have almost 4% more instructions

avikivity commented:
Hmm. I wonder if it's going to be worse after index caching is merged. 

What was the difference before the change back to switch? I'm curious how much an individual change costs.

wmitros commented:
I've found the main cause of bad performance in the cases with 4% degradation:
currently, when we're skipping a row (because is not in our range) we're returning from ```do_process_state()``` with the result of ```continuous_data_consumer::skip()```, even though ```skip()``` may return ```proceed::yes```, if we don't need to skip our entire buffer, causing the loop in ```continuous_data_consumer::process()``` to immediately call ```do_process_state()``` again.
This is more expensive, when ```do_process_state()``` is a coroutine.

If ```do_process_state()``` does not return when the result of ```skip()``` is  ```proceed::yes```, the biggest increases in instructions per fragment are about 2%, but the increase is in cases where skip = 0 (basically where we gain little to no advantage using more efficient row skips). (The same fix can be also applied to the existing version without coroutines. When compared to the fixed version, the increase is almost 3% in some cases, but it's still about 2% where we don't need to skip any rows using the coroutine)

The currently uploaded version reconstructs the generator object after every ```reset()``` call. If we change it to construct only the first time, and recognize the reset from within the coroutine, we can get rid of additional ~0.3% instructions, but to do that, we need to set a new flag in ```reset()```, and after (almost) every co_yield in the coroutine we have to add an if statement with a goto to the new state

avikivity commented:
Ok, I think this makes sense now. @tgrabiec @nyh please give an opinion on the end result (I see the patches need some folding).

avikivity commented on sstables/kl/reader_impl.hh:
```diff
@@ -190,224 +169,159 @@ public:
+                    }
+                    if (read_32(*_processing_data) != read_status::ready) {
+                        co_yield row_consumer::proceed::yes;
+                    }
```
Perhaps read_32 (and friends) and return a value, for which co_yield takes the following actions:
 - if the value indicates we can continue, the awaiter does nothing (await_suspend returns false)
 - if the value indicates we can't continue but should continue pushing data, does what it currently does for `yes`
 - similary for no

So we can write `co_yield read_32(*_processing_data);`

wmitros replied:
Nice idea, though to avoid the cost of suspending the coroutine, it will probably be ```await_ready``` instead of ```await_suspend``` (like in ```suspend_never```)

avikivity replied:
Yes.

avikivity commented on sstables/proceed_generator.hh:
```diff
@@ -0,0 +1,102 @@
+#include <seastar/core/coroutine.hh>
+#include "sstables/consumer.hh"
+
+class proceed_generator {
```
Please add documentation about the theory of operation of this thing.

wmitros replied:
Ok

wmitros replied:
I'm also not completely satisfied with this name, maybe we can come up with a better one.
I used "proceed" in the name because we can generate  ```data_consumer::proceed``` values with it.
You can say that the class is a generator, because you have function that yields values one by one, though the yielded values are actually only for controlling when to stop or whether to call primitive_consumer::consume(), and the actual mutation fragments are saved in the flat_mutation_reader::impl buffer.
It's kind of hard to create a general generator because you can customize it so much in c++(@avikivity  had an idea of [something similar](https://github.com/scylladb/scylla/pull/8653#issuecomment-849590716), but as far as I know, we didn't end up deciding on anything yet)

avikivity commented:
Please add a summary of the performance results to the cover letter, as well as a comparison of the number of lines of code in the main switch statement. So we can see that we're simplifying the logic in return for a small sacrifice in performance.

avikivity commented:
(the co_yield trick I described will lead to an even bigger reduction in line count).

nyh commented on sstables/mx/reader.cc:
```diff
@@ -403,9 +402,7 @@ class data_consume_rows_context_m : public data_consumer::continuous_data_consum
                 _state = state::RANGE_TOMBSTONE_MARKER;
-                goto range_tombstone_marker_label;
             // case state::RANGE_TOMBSTONE_MARKER:
-        range_tombstone_marker_label:
```
This patch does a lot of things (like here removing the gotos and labels which I wondered in the previous patch why you didn't remove them) that have nothing to do with the patch's description.

wmitros replied:
True, I will separate them in the final PR

nyh commented on sstables/mx/reader.cc:
```diff
@@ -350,350 +350,461 @@ class data_consume_rows_context_m : public data_consumer::continuous_data_consum
-                if (_consumer.consume_partition_end() == consumer_m::proceed::no) {
-                    co_yield consumer_m::proceed::no;
+        while (true) {
+            switch (_state) {
```
So, here, finally, is the missing switch I wondered about.
So how did this code work before this patch? I am guessing it didn't.
Patch series should be bisectable - this is important for "git bisect" but also to human reviewers like me who don't understand all these small incremental changes, which look wrong because they are (apparently) wrong - the previous patches were missing this while() and switch(), and because of that were wrong.

wmitros replied:
It's probable that the code didn't work for some patches, sorry for that. All patches in the new PR should work however.

nyh commented on sstables/mx/reader.cc:
```diff
@@ -903,7 +902,7 @@ class data_consume_rows_context_m : public data_consumer::continuous_data_consum
         , _column_translation(sst->get_column_translation(s, _header, sst->features()))
         , _has_shadowable_tombstones(sst->has_shadowable_tombstones())
-        , _gen(do_process_state(false))
+        , _gen(do_process_state())
```
The commit message says just "remove skip_partition". Why? Can you write a few additional words on why it was needed in the past and no longer is?

wmitros replied:
This will also appear only in the final version in the PR (the skip_partition was also added by me in an earlier commit in this series). The skip_partition was useful for checking if we called reset() and changed the state with that, but this can be done just by checking the state.

nyh commented on sstables/kl/reader_impl.hh:
```diff
@@ -159,8 +158,9 @@ private:
+    // assumes !primitive_consumer::active()
     bool non_consuming() const {
-        return !_consuming;
+        return false;
```
‎This _consuming thing was added in a previous patch, but now removed. So why add it at all?

wmitros replied:
Only after introducing the _consuming variable I realized that we only check non_consuming if !primitive_consumer::active(), and only then I realized that there are no states in kl reader's non_consuming() OR, where !primitive_consumer::active()

nyh commented on sstables/proceed_generator.hh:
```diff
@@ -24,24 +24,21 @@
 public:
     struct promise_type {
-        using handle_type = SEASTAR_INTERNAL_COROUTINE_NAMESPACE::coroutine_handle<promise_type>;
+        using handle_type = std::experimental::coroutine_handle<promise_type>;
```
Unless I'm missing something, you added this code in a previous patch. So please merge the fix into it.

wmitros replied:
I will avoid SEASTAR_INTERNAL_COROUTINE_NAMESPACE in the new PR altogether

nyh reviewed (changes requested):
This patch series is very impressive work, but unfortunately I cannot provide a meaningful review to it because I don't understand it. Through 20 or so patches, you take some very long pieces and make unexplained partial changes to pieces in the middle of this - changes which (if I understand correctly) leave non-working code, undo changes done in the previous patches.
For me, it would have been much easier to understand if you had one big patch **replacing** the entire code, so I can review the final working state, and not 20 non-working intermediate states. The last couple of patches can be about optimizing the already working code.

Alternatively, if the intermediate patches *are* working code, and *do* make sense, I was simply unable to understand how. You may need to either add a lot more text in the commit messages - explaining both the current patch and the plan for the next patches - or maybe you just need a smarter reviewer :-)

wmitros commented:
@tgrabiec  @nyh currently it's really hard/impossible to review the changes patch by patch, so you can only say how the overall change looks, or whether the performance loss is bad

I'm currently finishing dividing the patch into comprehensible commits, and I am planning to create a separate PR for it (this was supposed to be just an RFC)

wmitros commented on sstables/kl/reader_impl.hh:
```diff
@@ -190,224 +194,225 @@ public:
-                if (process(chunk) == row_consumer::proceed::no) {
-                    data.trim_front(i + chunk_size - chunk.size());
-                    return row_consumer::proceed::no;
+            // Testing hack: call process() for tiny chunks separately, to verify
```
Ok

wmitros commented on sstables/kl/reader_impl.hh:
```diff
@@ -170,6 +171,9 @@ private:
     bool _shadowable;
+
+    proceed_generator _gen;
+    temporary_buffer<char>* _processing_data;
```
In the original do_process_state method we had a reference to a buffer, which we modified inside the function. We can't really feed a variable to the coroutine, and we can't have a reference member for the buffer, because there isn't really any reasonable default value for it (before we run the coroutine), so I decided to have a bare pointer (we could have an optional reference but its basically the same thing)

wmitros commented on sstables/kl/reader_impl.hh:
```diff
@@ -190,224 +194,225 @@ public:
-                    return row_consumer::proceed::no;
+    }
+private:
+    proceed_generator do_process_state(bool skip_row) {
```
I will remove this bool in the final series, it meant to skip the entire case of ROW_START

wmitros commented on sstables/kl/reader_impl.hh:
```diff
@@ -190,224 +194,225 @@ public:
+            if (!skip_row) {
+            // case state::ROW_START:
+                if (read_short_length_bytes(*_processing_data, _key) != read_status::ready) {
+                    _state = state::DELETION_TIME;
```
The some states are still needed for the verify_end_state() method (we could make new bools or variables for it but I just left some of the states that we were using before)

The retrying is actually not added by this patch, you can find it in continuous_data_consumer::process(). Basically, when the function returns (or the coroutine yields) we recognize that primitive_consumer::active(), and we keep calling primitive_consumer::consume(data) until we can continue with our sstable reader

wmitros commented on sstables/mx/reader.cc:
```diff
@@ -404,6 +404,31 @@ class data_consume_rows_context_m : public data_consumer::continuous_data_consum
                 _state = state::RANGE_TOMBSTONE_MARKER;
                 goto range_tombstone_marker_label;
+            // case state::RANGE_TOMBSTONE_MARKER:
+        range_tombstone_marker_label:
```
At the time, I was thinking about removing all the labels next to gotos at once. In the PR we'll move and remove them together, one by one

wmitros commented:
I've rebased already, but I still have to test if the patch is bisectable, and write a cover letter, so I'll create the PR tomorrow

wmitros commented:
As I mentioned before, I'm closing this RFC to open a reviewable PR with a new cover letter
